src/sources/mithril.rs

```diff
@@ -0,0 +1,249 @@
+};
+use serde::Deserialize;
+use std::{path::Path, sync::Arc};
+use tracing::warn;
+
+use crate::framework::*;
+
+struct Feedback {
+    _multi: MultiProgress,
+    download_pb: ProgressBar,
+    validate_pb: ProgressBar,
+}
+
+impl Feedback {
+    fn indeterminate_progress_bar(owner: &mut MultiProgress) -> ProgressBar {
+        let pb = ProgressBar::new_spinner();
+
+        pb.set_style(
+            ProgressStyle::with_template("{spinner:.green} [{elapsed_precise}] {msg}").unwrap(),
+        );
+
+        owner.add(pb)
+    }
+
+    fn bytes_progress_bar(owner: &mut MultiProgress) -> ProgressBar {
+        let pb = ProgressBar::new_spinner();
+
+        pb.set_style(
+            ProgressStyle::with_template(
+                "{spinner:.green} [{elapsed_precise}] {bar:40.cyan/blue} {bytes}/{total_bytes} (eta: {eta}) {msg}",
+            )
+            .unwrap()
+            .progress_chars("#>-"),
+        );
+
+        owner.add(pb)
+    }
+}
+
+impl Default for Feedback {
+    fn default() -> Self {
+        let mut multi = MultiProgress::new();
+
+        Self {
+            download_pb: Self::bytes_progress_bar(&mut multi),
+            validate_pb: Self::indeterminate_progress_bar(&mut multi),
+            _multi: multi,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl mithril_client::feedback::FeedbackReceiver for Feedback {
+    async fn handle_event(&self, event: mithril_client::feedback::MithrilEvent) {
+        match event {
+            mithril_client::feedback::MithrilEvent::SnapshotDownloadStarted { .. } => {
+                self.download_pb.set_message("snapshot download started")
+            }
+            mithril_client::feedback::MithrilEvent::SnapshotDownloadProgress {
+                downloaded_bytes,
+                size,
+                ..
+            } => {
+                self.download_pb.set_length(size);
+                self.download_pb.set_position(downloaded_bytes);
+                self.download_pb.set_message("downloading Mithril snapshot");
+            }
+            mithril_client::feedback::MithrilEvent::SnapshotDownloadCompleted { .. } => {
+                self.download_pb.set_message("snapshot download completed");
+            }
+            mithril_client::feedback::MithrilEvent::CertificateChainValidationStarted {
+                ..
+            } => {
+                self.validate_pb
+                    .set_message("certificate chain validation started");
+            }
+            mithril_client::feedback::MithrilEvent::CertificateValidated {
+                certificate_hash: hash,
+                ..
+            } => {
+                self.validate_pb
+                    .set_message(format!("validating cert: {hash}"));
+            }
+            mithril_client::feedback::MithrilEvent::CertificateChainValidated { .. } => {
+                self.validate_pb.set_message("certificate chain validated");
+            }
+        }
+    }
```

Comment:
turn progress bar feedback into logs (try not to flood with download percentage logs)

Reply:
removed the progress bar and converted to logs and only displays download logs every 10 seconds